### PR TITLE
[WIP] Update project COSR logic to read historical hourly costs for assignments

### PR DESCRIPTION
**Pull request by joshmcclain**
**TLDR: We want to update COSR to use actual historical hourly costs, instead of average studio-wide costs. This is a WIP PR to get input on the direction I'm considering for this problem.**

The overall approach I'm using consists of:
- Creating a new `forecast_person_cost_windows` table. This is meant to contain historical snapshots of the effective hourly rate for each `ForecastPerson`. Right now I've set this up to be range-based, so any time someone's effective hourly cost changes (ie, skill tree band changes), we'll need to insert a new record into this table to reflect the new cost to the studio.
- Creating a new `CostOfServicesRenderedCalculator` utility class and passing it the necessary ingredients to calculate the daily costs for each `ForecastAssignment` during a given window of time.

Miscellaneous notes / caveats:
- This PR does *not* cover the actual insertion of records into the new `forecast_person_cost_windows` table. That happens out of band. I have some work done around this but I wanted to get some input on the larger direction before muddying the waters with that.
- This PR does *not* update any of the template logic to reflect these changes (eg, in the COSR explorer). Some changes there will be necessary because A) we'll no longer be able to show a topline cost/hr on the studio level like we do today and B) given the possibility that a person's cost/hr could change midway through the month, we'll need to potentially display multiple lines per person to reflect those different hourly costs. That's out of band for now.

I'll add additional notes inline explaining my thinking on various things.

### CODE REVIEW

**Comment by hhff** [comment]
Thank you Josh! Some thoughts (holding on reviewing until you're ready):

> This PR does not cover the actual insertion of records into the new forecast_person_cost_windows table. That happens out of band. I have some work done around this but I wanted to get some input on the larger direction before muddying the waters with that.

Feels like this could be done as a side effect to create/update to `FullTimePeriod` and the archiving of a `Finalization` (which marks a change in the individual's skill level and thus salary)

> This PR does not update any of the template logic to reflect these changes (eg, in the COSR explorer). Some changes there will be necessary because A) we'll no longer be able to show a topline cost/hr on the studio level like we do today and B) given the possibility that a person's cost/hr could change midway through the month, we'll need to potentially display multiple lines per person to reflect those different hourly costs. That's out of band for now.

Understood!

---

Some other thoughts:

1. This code won't handle the case when we're working with contractors. Usually we've agree'd on an hourly rate with our contractors, but that happens on the per-project level (it's possible a contractor is paid $80p/hr for Project A, and $100p/hr for Project B).
2. Additionally, because each month has variable amount of days, and thus a different amount of sellable hours, the cost per hour is variable in a given cost window in a given month... (oy)
3. When deriving this cost, I assume we'll use the numbers [here](https://github.com/sanctuarycomputer/stacks/blob/4afe0629a791ae72201a5556d1ece46454786020/app/models/review.rb#L54), but we should consider how this code will change if we make global updates to our salary scale (likely we need to version our salary scale by date)

COOL! Otherwise this all looks like it's in the right direction <3
